In [ ]:
# Import modules we will use
import pandas as pd
import numpy as np

# Read our data
sf_permits = pd.read_csv("Building_Permits.csv")

# Set seed for reproducibility
np.random.seed(0)

In [5]:
# look at a few rows of the nfl_data file. I can see a handful of missing data already!
sf_permits.sample(5)

Permit Number  Permit Type            Permit Type Definition  \
40553   201403039652            8            otc alterations permit   
169731  201510159735            3  additions alterations or repairs   
19180        M409787            8            otc alterations permit   
68047   201411191888            8            otc alterations permit   
64238        M527228            8            otc alterations permit   

       Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
40553            03/03/2014  3732  008            400                  NaN   
169731           10/15/2015  2609  028             79                  NaN   
19180            07/22/2013  4624  031            178                  NaN   
68047            11/19/2014  0039  109            294                  NaN   
64238            10/14/2014  1251  002            707                  NaN   

        Street Name Street Suffix  ...  Existing Construction Type  \
40553    Clementina            St  ...                         NaN   
169731  Buena Vista            Tr  ...                         5.0   
19180    West Point            Rd  ...                         NaN   
68047     Francisco            St  ...                         5.0   
64238          Cole            St  ...                         NaN   

       Existing Construction Type Description Proposed Construction Type  \
40553                                     NaN                        1.0   
169731                         wood frame (5)                        5.0   
19180                                     NaN                        NaN   
68047                          wood frame (5)                        5.0   
64238                                     NaN                        NaN   

       Proposed Construction Type Description Site Permit Supervisor District  \
40553                           constr type 1         NaN                 6.0   
169731                         wood frame (5)         NaN                 8.0   
19180                                     NaN         NaN                10.0   
68047                          wood frame (5)         NaN                 3.0   
64238                                     NaN         NaN                 5.0   

       Neighborhoods - Analysis Boundaries  Zipcode  \
40553                      South of Market  94103.0   
169731                 Castro/Upper Market  94117.0   
19180                Bayview Hunters Point  94124.0   
68047                          North Beach  94133.0   
64238                       Haight Ashbury  94117.0   

                                         Location      Record ID  
40553   (37.780460571778164, -122.40450626524974)  1334094491645  
169731   (37.76757916496494, -122.43793170417105)  1399356139170  
19180    (37.73524725436046, -122.38063828309745)  1311685491725  
68047   (37.805257822817126, -122.40998545760392)  1362881288870  
64238    (37.76836885973765, -122.45074431487859)   135886493776  

[5 rows x 43 columns]

## See how many missing data points we have

In [10]:
# Get the number of missing data points per column
missing_values_count = sf_permits.isnull().sum()

# Look at the number of missing points in the data
missing_values_count[0:]

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                      196684
Street Name                                    0
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing S

In [13]:
# How many total missing values do we have?
total_cells = np.product(sf_permits.shape)
total_missing = missing_values_count.sum()

# Percent of data that is missing
(total_missing/total_cells) * 100

26.26002315058403

As we can see, more than a quarter of our data corresponds to non-data values. Now we will analyze the data in order to know the reason for these null values.

## Figure out why the data is missing

To deal with missing values we need to resort to intuition to know why those values are missing. A fundamental question we must ask ourselves is:
#### Is this value missing because it was not recorded or because it does not exist?
If a value is missing because it does not exist (such as the age of the children of a family that has no children), there is no point in guessing what value it might be.
If a value is missing because it was not recorded, we can guess what it might have been based on the other values in that column and row. (This is known as "IMPUTATION" and we will see it below)

In [14]:
# Look at the number of missing points in the data
missing_values_count[0:]

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                      196684
Street Name                                    0
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing S

As we can see, we have many missing values in variables such as:

First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit and Permit Expiration Date

These variables refer to situations over time. For example, the variable "Structural Notification" has 191978 which may refer to the fact that said notification has not yet arrived. Therefore, it is data that does not exist and is not related to poor data collection.

If you're doing very careful data analysis, this is the point at which you'd look at each column individually to figure out the best strategy for filling those missing values. For the rest of this notebook, we'll cover some "quick and dirty" techniques that can help you with missing values but will probably also end up removing some useful information or adding some noise to your data.

## Drop missing values

If you're in a hurry or don't have a reason to figure out why your values are missing, one option you have is to just remove any rows or columns that contain missing values.

In [15]:
# Remove all the rows that contain a missing value
sf_permits.dropna()

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location, Record ID]
Index: []

[0 rows x 43 columns]

This is because every row in our dataset had at least one missing value. We might have better luck removing all the columns that have at least one missing value instead.

In [18]:
# Remove all columns with at least one missing value
columns_with_na_dropped = sf_permits.dropna(axis=1)
columns_with_na_dropped.head()

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Name Current Status  \
0           05/06/2015  0326  023            140       Ellis        expired   
1           04/19/2016  0306  007            440       Geary         issued   
2           05/27/2016  0595  203           1647     Pacific      withdrawn   
3           11/07/2016  0156  011           1230     Pacific       complete   
4           11/28/2016  0342  001            950      Market         issued   

  Current Status Date  Filed Date      Record ID  
0          12/21/2017  05/06/2015  1380611233945  
1          08/03/2017  04/19/2016  1420164406718  
2          09/26/2017  05/27/2016  1424856504716  
3          07/24/2017  11/07/2016  1443574295566  
4          12/01/2017  11/28/2016   144548169992

In [21]:
# Just how much data did we lose?
print("Columns in original dataset: %d \n" % sf_permits.shape[1])
print("Columns in original dataset: %d \n" % columns_with_na_dropped.shape[1])

Columns in original dataset: 43 

Columns in original dataset: 12 



## Filling in missing values automatically


In [25]:
# Get a small subset of the NFL dataset
subset_sf_permits = sf_permits.loc[:, 'Street Number Suffix':'Location'].head()
subset_sf_permits

Street Number Suffix Street Name Street Suffix  Unit Unit Suffix  \
0                  NaN       Ellis            St   NaN         NaN   
1                  NaN       Geary            St   0.0         NaN   
2                  NaN     Pacific            Av   NaN         NaN   
3                  NaN     Pacific            Av   0.0         NaN   
4                  NaN      Market            St   NaN         NaN   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date  ... TIDF Compliance  \
0          12/21/2017  05/06/2015  11/09/2015  ...             NaN   
1          08/03/2017  04/19/2016  08/03/2017  ...             NaN   
2          09/26/2017  05/27/2016         NaN  ...             NaN   
3          07/24/2017  11/07/2016  07/18/2017  ...             NaN   
4          12/01/2017  11/28/2016  12/01/2017  ...             NaN   

  Existing Construction Type Existing Construction Type Description  \
0                        3.0                          constr type 3   
1                        3.0                          constr type 3   
2                        1.0                          constr type 1   
3                        5.0                         wood frame (5)   
4                        3.0                          constr type 3   

   Proposed Construction Type  Proposed Construction Type Description  \
0                         NaN                                     NaN   
1                         NaN                                     NaN   
2                         1.0                           constr type 1   
3                         5.0                          wood frame (5)   
4                         NaN                                     NaN   

  Site Permit Supervisor District Neighborhoods - Analysis Boundaries  \
0         NaN                 3.0                          Tenderloin   
1         NaN                 3.0                          Tenderloin   
2         NaN                 3.0                        Russian Hill   
3         NaN                 3.0                            Nob Hill   
4         NaN                 6.0                          Tenderloin   

   Zipcode                                   Location  
0  94102.0  (37.785719256680785, -122.40852313194863)  
1  94102.0   (37.78733980600732, -122.41063199757738)  
2  94109.0    (37.7946573324287, -122.42232562979227)  
3  94109.0   (37.79595867909168, -122.41557405519474)  
4  94102.0   (37.78315261897309, -122.40950883997789)  

[5 rows x 35 columns]

In [26]:
# Replace all NA's with 0
subset_sf_permits.fillna(0)

C:\Users\usuario\AppData\Local\Temp\ipykernel_10356\1037732806.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  subset_sf_permits.fillna(0)


Street Number Suffix Street Name Street Suffix  Unit  Unit Suffix  \
0                     0       Ellis            St   0.0            0   
1                     0       Geary            St   0.0            0   
2                     0     Pacific            Av   0.0            0   
3                     0     Pacific            Av   0.0            0   
4                     0      Market            St   0.0            0   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date  ... TIDF Compliance  \
0          12/21/2017  05/06/2015  11/09/2015  ...               0   
1          08/03/2017  04/19/2016  08/03/2017  ...               0   
2          09/26/2017  05/27/2016           0  ...               0   
3          07/24/2017  11/07/2016  07/18/2017  ...               0   
4          12/01/2017  11/28/2016  12/01/2017  ...               0   

  Existing Construction Type  Existing Construction Type Description  \
0                        3.0                           constr type 3   
1                        3.0                           constr type 3   
2                        1.0                           constr type 1   
3                        5.0                          wood frame (5)   
4                        3.0                           constr type 3   

   Proposed Construction Type  Proposed Construction Type Description  \
0                         0.0                                       0   
1                         0.0                                       0   
2                         1.0                           constr type 1   
3                         5.0                          wood frame (5)   
4                         0.0                                       0   

   Site Permit  Supervisor District Neighborhoods - Analysis Boundaries  \
0            0                  3.0                          Tenderloin   
1            0                  3.0                          Tenderloin   
2            0                  3.0                        Russian Hill   
3            0                  3.0                            Nob Hill   
4            0                  6.0                          Tenderloin   

   Zipcode                                   Location  
0  94102.0  (37.785719256680785, -122.40852313194863)  
1  94102.0   (37.78733980600732, -122.41063199757738)  
2  94109.0    (37.7946573324287, -122.42232562979227)  
3  94109.0   (37.79595867909168, -122.41557405519474)  
4  94102.0   (37.78315261897309, -122.40950883997789)  

[5 rows x 35 columns]

I could also be a bit more savvy and replace missing values with whatever value comes directly after it in the same column. (This makes a lot of sense for datasets where the observations have some sort of logical order to them.)

In [27]:
# Replace all NA's the value that comes directly after it in the same column, 
# Then replace all the reamining na's with 0
subset_sf_permits.fillna(method="bfill",axis=0).fillna(0)

C:\Users\usuario\AppData\Local\Temp\ipykernel_10356\677563574.py:3: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  subset_sf_permits.fillna(method="bfill",axis=0).fillna(0)
C:\Users\usuario\AppData\Local\Temp\ipykernel_10356\677563574.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  subset_sf_permits.fillna(method="bfill",axis=0).fillna(0)


Street Number Suffix Street Name Street Suffix  Unit  Unit Suffix  \
0                   0.0       Ellis            St   0.0          0.0   
1                   0.0       Geary            St   0.0          0.0   
2                   0.0     Pacific            Av   0.0          0.0   
3                   0.0     Pacific            Av   0.0          0.0   
4                   0.0      Market            St   0.0          0.0   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date  ... TIDF Compliance  \
0          12/21/2017  05/06/2015  11/09/2015  ...             0.0   
1          08/03/2017  04/19/2016  08/03/2017  ...             0.0   
2          09/26/2017  05/27/2016  07/18/2017  ...             0.0   
3          07/24/2017  11/07/2016  07/18/2017  ...             0.0   
4          12/01/2017  11/28/2016  12/01/2017  ...             0.0   

  Existing Construction Type  Existing Construction Type Description  \
0                        3.0                           constr type 3   
1                        3.0                           constr type 3   
2                        1.0                           constr type 1   
3                        5.0                          wood frame (5)   
4                        3.0                           constr type 3   

   Proposed Construction Type  Proposed Construction Type Description  \
0                         1.0                           constr type 1   
1                         1.0                           constr type 1   
2                         1.0                           constr type 1   
3                         5.0                          wood frame (5)   
4                         0.0                                       0   

   Site Permit  Supervisor District Neighborhoods - Analysis Boundaries  \
0          0.0                  3.0                          Tenderloin   
1          0.0                  3.0                          Tenderloin   
2          0.0                  3.0                        Russian Hill   
3          0.0                  3.0                            Nob Hill   
4          0.0                  6.0                          Tenderloin   

   Zipcode                                   Location  
0  94102.0  (37.785719256680785, -122.40852313194863)  
1  94102.0   (37.78733980600732, -122.41063199757738)  
2  94109.0    (37.7946573324287, -122.42232562979227)  
3  94109.0   (37.79595867909168, -122.41557405519474)  
4  94102.0   (37.78315261897309, -122.40950883997789)  

[5 rows x 35 columns]